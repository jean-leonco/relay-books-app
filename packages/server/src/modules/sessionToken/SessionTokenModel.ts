import mongoose, { Document, Model, Types } from 'mongoose';

import { PLATFORM } from '../../common/utils';
import { isActiveMongooseField } from '../../core/mongoose/withMongooseFields';

const { ObjectId } = mongoose.Schema.Types;

export enum SESSION_TOKEN_SCOPES {
  RESET_PASSWORD = 'BOOKSAPP:RESET_PASSWORD', // used when user requested a password reset
  SESSION = 'BOOKSAPP:SESSION', // token for a valid user session
}

const Schema = new mongoose.Schema(
  {
    user: {
      type: ObjectId,
      ref: 'User',
      description: 'User owner of this token.',
    },
    ip: {
      type: String,
      description: 'Agent IP that came from request headers.',
      required: true,
      index: true,
    },
    channel: {
      type: String,
      required: true,
      default: PLATFORM.APP,
      description: 'Channel that this token belongs too. ex: APP',
      enum: Object.values(PLATFORM),
    },
    name: {
      type: String,
      description: 'Name of this token. ex: Jean Linux',
    },
    scope: {
      type: String,
      description: 'Scope of this session token. ex: SESSION',
    },
    expiresIn: {
      type: Date,
      description: 'When this session token expires.',
    },
    isBlocked: {
      type: Boolean,
      description: 'If the token is blocked.',
      default: false,
      required: true,
      index: true,
    },
    ...isActiveMongooseField,
  },
  {
    timestamps: {
      createdAt: 'createdAt',
      updatedAt: 'updatedAt',
    },
    collection: 'SessionToken',
  },
);

export interface ISessionToken extends Document {
  user: Types.ObjectId;
  ip: string;
  channel: string;
  isActive: boolean;
  isBlocked: boolean;
  name: string;
  scope: string;
  expiresIn: Date | null;
  createdAt: Date;
  updatedAt: Date;
}

const SessionTokenModel: Model<ISessionToken> = mongoose.model('SessionToken', Schema);

export default SessionTokenModel;
