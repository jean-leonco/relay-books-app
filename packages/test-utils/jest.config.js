const pkg = require('./package');

module.exports = {
  rootDir: './',
  name: pkg.name,
  displayName: pkg.name.toUpperCase(),
  testPathIgnorePatterns: ['/node_modules/', './build'],
  coverageReporters: ['lcov', 'html'],
  setupFilesAfterEnv: ['<rootDir>/test/setupTestFramework.js'],
  globalSetup: '<rootDir>/test/setup.js',
  globalTeardown: '<rootDir>/test/teardown.js',
  transform: {
    // '^.+\\.(js|jsx|ts|tsx)?$': require.resolve('babel-jest'),
    '^.+\\.(js|jsx|ts|tsx)?$': '<rootDir>/test/babel-transformer',
  },
};
