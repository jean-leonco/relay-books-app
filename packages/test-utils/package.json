{
  "name": "@booksapp/test-utils",
  "version": "0.0.1",
  "private": true,
  "license": "UNLICENSED",
  "main": "src/index.ts",
  "module": "src/index.ts",
  "scripts": {
    "test": "jest",
    "test:watch": "jest --watch"
  },
  "dependencies": {
    "@booksapp/babel": "^0.0.1",
    "@booksapp/types": "^0.0.1",
    "@koa/router": "^9.3.1",
    "babel-jest": "^26.1.0",
    "graphql-relay": "0.6.0",
    "koa": "^2.13.0",
    "koa-bodyparser": "^4.3.0",
    "mongoose": "^5.10.7",
    "supertest": "^5.0.0"
  }
}
