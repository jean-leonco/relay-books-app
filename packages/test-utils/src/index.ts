export { defaultFrozenKeys, sanitizeValue, sanitizeTestObject } from './objectNormalization';
export { httpRequestGraphql } from './httpRequestGraphql';
