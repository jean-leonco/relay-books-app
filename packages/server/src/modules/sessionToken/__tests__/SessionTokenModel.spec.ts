import { sanitizeTestObject } from '@booksapp/test-utils';

import {
  clearDbAndRestartCounters,
  connectMongoose,
  createSessionToken,
  disconnectMongoose,
} from '../../../../test/helpers';

import SessionTokenModel from '../SessionTokenModel';

beforeAll(connectMongoose);

beforeEach(clearDbAndRestartCounters);

afterAll(disconnectMongoose);

describe('SessionTokenModel', () => {
  it('should create as many session token records as needed', async () => {
    // creating 5 session tokens with same ip, channel and scope
    await createSessionToken();
    await createSessionToken();
    await createSessionToken();
    await createSessionToken();
    await createSessionToken();

    await SessionTokenModel.createIndexes();

    const sessionTokens = await SessionTokenModel.find().lean();

    expect(sessionTokens.length).toBe(5);
    expect(sanitizeTestObject(sessionTokens)).toMatchSnapshot();
  });
});
