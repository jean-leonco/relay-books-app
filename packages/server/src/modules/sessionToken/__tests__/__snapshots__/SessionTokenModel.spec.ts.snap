// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SessionTokenModel should create as many session token records as needed 1`] = `
Object {
  "0": Object {
    "__v": 0,
    "_id": "ObjectId",
    "channel": "APP",
    "createdAt": "FROZEN-CREATEDAT",
    "ip": "::ffff:127.0.0.1",
    "isActive": true,
    "isBlocked": false,
    "scope": "BOOKSAPP:SESSION",
    "updatedAt": "FROZEN-UPDATEDAT",
  },
  "1": Object {
    "__v": 0,
    "_id": "ObjectId",
    "channel": "APP",
    "createdAt": "FROZEN-CREATEDAT",
    "ip": "::ffff:127.0.0.1",
    "isActive": true,
    "isBlocked": false,
    "scope": "BOOKSAPP:SESSION",
    "updatedAt": "FROZEN-UPDATEDAT",
  },
  "2": Object {
    "__v": 0,
    "_id": "ObjectId",
    "channel": "APP",
    "createdAt": "FROZEN-CREATEDAT",
    "ip": "::ffff:127.0.0.1",
    "isActive": true,
    "isBlocked": false,
    "scope": "BOOKSAPP:SESSION",
    "updatedAt": "FROZEN-UPDATEDAT",
  },
  "3": Object {
    "__v": 0,
    "_id": "ObjectId",
    "channel": "APP",
    "createdAt": "FROZEN-CREATEDAT",
    "ip": "::ffff:127.0.0.1",
    "isActive": true,
    "isBlocked": false,
    "scope": "BOOKSAPP:SESSION",
    "updatedAt": "FROZEN-UPDATEDAT",
  },
  "4": Object {
    "__v": 0,
    "_id": "ObjectId",
    "channel": "APP",
    "createdAt": "FROZEN-CREATEDAT",
    "ip": "::ffff:127.0.0.1",
    "isActive": true,
    "isBlocked": false,
    "scope": "BOOKSAPP:SESSION",
    "updatedAt": "FROZEN-UPDATEDAT",
  },
}
`;
